### DISCUSSION TITLE
How to draw toy MC?

### DISCUSSION THREAD

**Discussion by zjkjsd** [scikit-hep/pyhf#2235] [Q&A]
Dear experts,

Could you please assist me in understanding how to use pyhf for drawing toy Monte Carlo (MC) samples? I came across the following code snippet, which seems to draw toys from the template model:
```
toys = model.make_pdf(pyhf.tensorlib.astensor(np.asarray(model.config.suggested_init()))).sample((n_toys,))
```
1. My understanding is that we first sum all the components and then draw samples from the Poisson distribution defined for each bin, with the mean set to the count in the corresponding template bin. Is this understanding correct? And is this the correct approach for drawing toys?

2. Furthermore, if I wish to draw toys with scaling factors applied to the signal and background components (for generating validation plots like linearity plots), should I replace model.config.suggested_init() with a list of the scaling factors?

3. If my understanding is correct, would the following code also achieve the desired outcome? (I included it to confirm my understanding.)
```
@joblib.delayed
def compute_toys(sig_hist, bkg_hist, sig_scale, bkg_scale, itry):
    MC = sig_scale * sig_hist + bkg_scale * bkg_hist
    rng = np.random.default_rng(itry)
    toy = rng.poisson(MC)
    return toy
ntoys=20000
toys = joblib.Parallel(-1)(compute_toys(sig_hist, bkg_hist, 0.5, 0.8, i) for i in range(ntoys))
```
Thank you for your support!
Boyang

**Comment by kratsg** [comment]
If the question is about how to throw toys using `pyhf`, we have calculators that do this for you. In particular, see [this example](https://scikit-hep.org/pyhf/examples/notebooks/learn/UsingCalculators.html#Toy-Based) in the documentation as well as [this example](https://scikit-hep.org/pyhf/examples/notebooks/toys.html) reproducing a key plot in the HistFactory paper.

If you're asking about how to do it yourself, look here: https://github.com/scikit-hep/pyhf/blob/c096b33cf859246771e06099ee1b48063ca051cd/src/pyhf/infer/calculators.py#L756-L779 . In particular, you need to evaluate the model at the fitted MLE and use the $\hat{\theta}$ parameters to initialize your pdf to sample from (rather than the default `suggested_init`). At that point, we just sample as you typically expect from the "scipy API" in the same way.

1. Yes, but it's Poisson *and* Gaussian.
2. Yes. But see my description/explanation above.
3. That code seems to be more generic about using a Poisson to make toys.

**Reply by alexander-held** [reply]
The key thing missing in your example @zjkjsd are the fluctuations of auxiliary data ("global observables"). For nuisance parameters with Gaussian constraints, the corresponding auxiliary data is drawn from Gaussians (that's where the Gaussian in @kratsg's 1. appears). If you look at the output of `pyhf`'s implementation, you will see that both the counts per bin and the auxiliary data fluctuate.

**Comment by zjkjsd** [comment]
Hi Giordon and Alex,

Thank you for your help, and I have a question for the code below:
```
# define samples and parameters
signal = 10000   # single bin
background = 0 # turned off for simplicity
background_uncertainty = 0
model = pyhf.simplemodels.uncorrelated_background(
    [signal], [background], [background_uncertainty]
)
# define toy par and generate toys
toy_pars = [0.1, 0] 
pdf_toy = model.make_pdf(pyhf.tensorlib.astensor(toy_pars))
toys = pdf_toy.sample((1000,))

# calculate the standard deviation of the actual data over all toys
sd = np.std(toys[:,0])
print(sd)
```
I would expect the standard deviation `sd` to be around `0.1 * sqrt(10000)=10`, but the result from the code above is `31.6`, why is that? It seems to be doing `sqrt( 0.1 * 10000)`, but I thought this was incorrect.

**Reply by alexander-held** [reply]
The fluctuations are taken from a Poisson with a rate corresponding to the yield at which you initialize. In your case `toy_pars` sets the yield via `model.make_pdf` to `0.1*10000`. The fluctuations around that number have a standard deviation of `sqrt(0.1*10000)`.

**Reply by zjkjsd** [reply]
Thank you!

**Reply by zjkjsd** [reply]
A follow up question: if my model (still a single bin) now has 3 fitting parameters, `mu_sig, mu_bkg, and uncorr_bkg`, I want to draw toys with a yield still corresponding to `0.1`, how should I initialize it? It would probably be `0.1` for `mu_sig and mu_bkg`, but not sure for `uncorr_bkg`. Thanks again for your quick response!

**Reply by alexander-held** [reply]
If `uncorr_bkg` is a `shapesys` modifier (like in the case of `simplemodels.uncorrelated_background`) then this comes with a Poisson constraint and a default parameter value of `1`. Unless you have a specific reason to change this I would expect that you generally would want to use that value.

**Reply by zjkjsd** [reply]
Thank you, Alex! I asked if I should change the default value because I wanted to draw toys with a yield of `0.1` and I was not sure how to set things up. My general question is: for toy generation, what values should be used to initialize the parameters that are not from the `normfactor` (e.g. `shapesys`, `staterror`, `histosys` ...). In the example above where I want to draw toys with a yield of `0.1`:
1. should I set all `normfactor` to `0.1`?
2. should I set all other modifiers to `1`? How to understand this?
3. In the `shapesys` case, you mentioned its poisson constraint, is it the reason why we keep it `1`?

I appreciate your help and time!

**Reply by alexander-held** [reply]
I'm afraid I cannot give you a general answer, this depends on what you want to do with the toys. In a setup based on Asimov data without reference to any observed data you are free to set `normfactor` modifiers to whatever you like (depends on what you want to achieve) and I would generally expect that you would want to set other parameters to values that maximize the associated constraint terms. I would recommend `cabinetry` for this, specifically [`cabinetry.model_utils.asimov_parameters`](https://cabinetry.readthedocs.io/en/latest/api.html#cabinetry.model_utils.asimov_parameters). The `shapesys` value of `1` is to achieve that, similarly Gaussian-constrained parameters like `normsys` / `histosys` would be initialized at `0`.

**Reply by zjkjsd** [reply]
Thank you very much! I am trying to do a linearity test where I generate toys with varying signal strength (`normfactor`) from `0.1 to 2`, and use my nominal workspace to fit the toys and check if I get the correct signal strength value. I am not sure if this is an Asimov test, but do you think the cabinetry function `asimov_parameters` works for this? Reaaaly appreciate it!!

**Reply by alexander-held** [reply]
This sounds like something you would do with just simulated data and with no reference to observed data so yes, that function is appropriate.

**Reply by zjkjsd** [reply]
Thank you, Alex! Yeah, I don't use observed data in the linearity test. Could you please confirm the steps below are reasonable procedures according to our discussions?
1. run `toy_pars = asimov_parameters(model)`
2. change the normfactor parameters according to my linearity test, e.g. `toy_pars[:2] = [0.1, 0.1]`
3. generate toys with `model.make_pdf(pyhf.tensorlib.astensor(toy_pars)).sample((ntoys,))`

I am trying to understand why `asimov_parameters` should be used for my linearity test, i.e. why `we want to set other parameters to values that maximize the associated constraint terms`. Now I guess that: varying the signal strength does not change our understanding of the systematics, so we should use our best knowledge in the constraint terms, which correspond to 1 for `shapesys` and 0 for `normsys`/`histosys`. Is this correct? Thanks again!

**Reply by alexander-held** [reply]
I do not know what you are ultimately trying to do and why you are generating toys. I expect you have a specific reason to generate them and in that case the choice of parameters would be motivated by what you are aiming to do with the toys. Generically all I can say is that in the absence of any additional knowledge (like from incorporating information from observed data), the most sensible setting for the parameters is the one maximizing the constraint terms and that holds true independent of any normalization factor values as you say.

I am not sure what linearity you are looking for and why you need toys for that. If you want to make sure you fit the injected normalization factor value again then this is something that happens by design (in the absence of technical problems) and it is unclear to me which additional information toys provide.

**Reply by zjkjsd** [reply]
Thank you for the explanation! Again, this is very helpful! I have 2 goals for generating toys:
1. fit the injected normalization factor value
2. calculate the uncertainty of the fitted normalization factors

It's good to know some happen by design, thanks! I am trying to do this: I have 2 workspaces whose background samples are constructed from different regions, `w_sig` from the signal region and `w_sb` from sidebands. I am trying to generate toys with my `w_sig` workspace and fit toys with `w_sb`. This way, I can estimate how well the sidebands describe the signal region, and if the sidebands are good enough (i.e. doesn't bias the signal strength). Does it make sense to use toys for my goals?

**Reply by alexander-held** [reply]
In general I would say that 1) happens by design and for 2) you can also look at the uncertainty in your fit to the Asimov dataset which might already tell you what you are after. See https://arxiv.org/abs/1007.1727 for more about this. We commonly use toys in cases where the asymptotic results derived in that paper do not hold. Another common use case is the estimation of trial factors for the look elsewhere effect. There are applications beyond that but I would say they are more specific and usually more rare.

I probably do not understand the setup you have. What sometimes happens is that the background estimate for a signal region is obtained from sidebands which you might fit first in a separate step or simultaneously with your signal region. I do not spontaneously see a need for toys in such a setup. I would recommend chatting with colleagues from Belle II which might be more familiar with your setup and can perhaps suggest common approaches used for such cases.

**Reply by zjkjsd** [reply]
Thanks for the explanation and suggestions! I will talk with my advisor and other Belle II people. To clarify my use of sidebands, I am taking the sidebands as a single sample to model a bkg component (FakeD) in the signal region, this way I don't need to use MC for this bkg (potential Data/MC discrepancy). I have heard fitting signal region and sidebands simultaneously, but I am not sure how to do it for my analysis. It sounds like I need to have templates for FakeD in both regions, and I am not sure how to avoid using MC.

**Reply by zjkjsd** [reply]
> I'm afraid I cannot give you a general answer, this depends on what you want to do with the toys. In a setup based on Asimov data without reference to any observed data you are free to set `normfactor` modifiers to whatever you like (depends on what you want to achieve) and I would generally expect that you would want to set other parameters to values that maximize the associated constraint terms. I would recommend `cabinetry` for this, specifically [`cabinetry.model_utils.asimov_parameters`](https://cabinetry.readthedocs.io/en/latest/api.html#cabinetry.model_utils.asimov_parameters). The `shapesys` value of `1` is to achieve that, similarly Gaussian-constrained parameters like `normsys` / `histosys` would be initialized at `0`.

Hi @alexander-held Alex,

Could you please comment on why `staterror` has initial value at 1 instead of 0, though it is a gaussian constrained parameter? Thank you very much! 

Edit: I think I understand a bit more now. It is because staterror is a multiplicative modifier and normsys/histosys are additive modifiers, that's why their initial values are set differently, though both maximize the prior probability. Is this correct?

**Reply by alexander-held** [reply]
Yes, I think your interpretation is correct. Equation 29 in https://cds.cern.ch/record/1456844 shows how to relate two nuisance parameter representations, $\alpha$ and $\beta$. For Gaussian-constrained parameters I think about them as $\alpha$, with auxiliary data 0 (so nuisance parameter = 0 maximizes constraint term). This choice of parameter standardization is just a convention. For Poisson-constrained parameters I think about them in this $\beta$ representation. I think the `pyhf` implementation for `staterror` follows the ROOT implementation, where one can switch between Poisson and Gaussian constraint, but the parameter value of 1 maximizes the constraint term in both cases. I assume what happens is an internal reparameterization in ROOT to make this work. In `pyhf`, where `staterror` currently can only be a Gaussian, that same reparameterization presumably also happens. cc @kratsg

**Reply by alexander-held** [reply]
Some more details: `staterror` is a bit special in that sense (I think caused by the feature that in ROOT you can switch constraint term for it). For `normsys` and `histosys` with Gaussian constraints the auxiliary data is 0.

> normsys/histosys are additive modifiers

That is correct for `histosys`, but `normsys` is multiplicative. For the `normsys` case you have a nuisance parameter, which then through an inter-/extrapolation function results in a factor that is multiplied by the sample it acts on.

**Reply by zjkjsd** [reply]
Thank you for clarifying! The `normsys` and `staterror` are more complicated than I thought!

>  In `pyhf`, where `staterror` currently can only be a Poisson, that same reparameterization presumably also happens. cc @kratsg

1. Do you mean `staterror currently can only be a Gauss`? It is listed as a gaussian constrained modifier from the [pyhf documentation page](https://pyhf.readthedocs.io/en/v0.7.6/intro.html). I understand your point though!

2. `normsys` and `normfactor` are both multiplicative modifiers, but the fit returned values probably should be interpreted differently, right? For example, if a fitted `normfactor` is 1.1, this means `fitted / model_expectation = 1.1`. For a `normsys`, it would mean `fitted - model_expectation = 1.1 * normsys_input`. Is my understanding correct?

**Reply by alexander-held** [reply]
1. Yes, `staterror` has a Gaussian constraint in `pyhf`, sorry for the mixup. I'll edit the comment above to prevent confusion.
2. For `normfactor`: yes, you're right. For `normsys` the exact form depends on the interpolation function as well. The model prediction is a function that takes both the nuisance parameter value (`1.1` in your case), the nominal sample yield (= what you get with nuisance parameter value of `0`), and the sample yields in the up/down variations.
